File: src/main/kotlin/samlang/interpreter/ExpressionInterpreter.kt
package samlang.interpreter

import kotlinx.collections.immutable.plus
import samlang.ast.common.BinaryOperator.AND
import samlang.ast.common.BinaryOperator.DIV
import samlang.ast.common.BinaryOperator.EQ
import samlang.ast.common.BinaryOperator.GE
import samlang.ast.common.BinaryOperator.GT
import samlang.ast.common.BinaryOperator.LE
import samlang.ast.common.BinaryOperator.LT
import samlang.ast.common.BinaryOperator.MINUS
import samlang.ast.common.BinaryOperator.MOD
import samlang.ast.common.BinaryOperator.MUL
import samlang.ast.common.BinaryOperator.NE
import samlang.ast.common.BinaryOperator.OR
import samlang.ast.common.BinaryOperator.PLUS
import samlang.ast.lang.CheckedExprVisitor
import samlang.ast.lang.Expression
import samlang.ast.lang.Expression.Binary
import samlang.ast.lang.Expression.FieldAccess
import samlang.ast.lang.Expression.FunctionApplication
import samlang.ast.lang.Expression.IfElse
import samlang.ast.lang.Expression.Lambda
import samlang.ast.lang.Expression.Literal
import samlang.ast.lang.Expression.Match
import samlang.ast.lang.Expression.MethodAccess
import samlang.ast.lang.Expression.ModuleMember
import samlang.ast.lang.Expression.ObjectConstructor
import samlang.ast.lang.Expression.Panic
import samlang.ast.lang.Expression.This
import samlang.ast.lang.Expression.TupleConstructor
import samlang.ast.lang.Expression.Unary
import samlang.ast.lang.Expression.Val
import samlang.ast.lang.Expression.Variable
import samlang.ast.lang.Expression.VariantConstructor
import samlang.ast.common.Literal.BoolLiteral
import samlang.ast.common.Literal.IntLiteral
import samlang.ast.common.Literal.StringLiteral
import samlang.ast.common.Literal.UnitLiteral
import samlang.ast.lang.Pattern
import samlang.ast.lang.Type
import samlang.ast.lang.UnaryOperator

internal object ExpressionInterpreter : CheckedExprVisitor<InterpretationContext, Value> {

    private fun blameTypeChecker(): Nothing = error(message = "Bad type checker!")

    fun eval(expression: Expression, context: InterpretationContext): Value =
        expression.accept(visitor = ExpressionInterpreter, context = context)

    override fun visit(expression: Literal, context: InterpretationContext): Value = when (val l = expression.literal) {
        UnitLiteral -> Value.UnitValue
        is IntLiteral -> Value.IntValue(value = l.value)
        is StringLiteral -> Value.StringValue(value = l.value)
        is BoolLiteral -> Value.BoolValue(value = l.value)
    }

    override fun visit(expression: This, context: InterpretationContext): Value =
        context.localValues["this"] ?: blameTypeChecker()

    override fun visit(expression: Variable, context: InterpretationContext): Value =
        context.localValues[expression.name] ?: blameTypeChecker()

    override fun visit(expression: ModuleMember, context: InterpretationContext): Value =
        context.classes[expression.moduleName]?.functions?.get(key = expression.memberName) ?: blameTypeChecker()

    override fun visit(expression: TupleConstructor, context: InterpretationContext): Value.TupleValue =
        Value.TupleValue(tupleContent = expression.expressionList.map { eval(expression = it, context = context) })

    override fun visit(expression: ObjectConstructor, context: InterpretationContext): Value.ObjectValue {
        val objectContent = mutableMapOf<String, Value>()
        expression.spreadExpression?.let { e ->
            val obj = eval(expression = e, context = context) as Value.ObjectValue
            objectContent.putAll(from = obj.objectContent)
        }
        expression.fieldDeclarations.forEach { declaration ->
            when (declaration) {
                is ObjectConstructor.FieldConstructor.Field -> {
                    objectContent[declaration.name] = eval(expression = declaration.expression, context = context)
                }
                is ObjectConstructor.FieldConstructor.FieldShorthand -> {
                    objectContent[declaration.name] = eval(
                        expression = Variable(
                            range = declaration.range,
                            type = declaration.type,
                            name = declaration.name
                        ),
                        context = context
                    )
                }
            }
        }
        return Value.ObjectValue(objectContent = objectContent)
    }

    override fun visit(expression: VariantConstructor, context: InterpretationContext): Value.VariantValue =
        Value.VariantValue(tag = expression.tag, data = eval(expression = expression.data, context = context))

    override fun visit(expression: FieldAccess, context: InterpretationContext): Value {
        val thisValue = eval(expression = expression.expression, context = context) as Value.ObjectValue
        return thisValue.objectContent[expression.fieldName] ?: blameTypeChecker()
    }

    override fun visit(expression: MethodAccess, context: InterpretationContext): Value.FunctionValue {
        val (id, _) = expression.expression.type as Type.IdentifierType
        val thisValue = eval(expression = expression.expression, context = context)
        val methodValue = context.classes[id]?.methods?.get(key = expression.methodName) ?: blameTypeChecker()
        val newCtx = context.copy(localValues = context.localValues.plus(pair = "this" to thisValue))
        methodValue.context = newCtx
        return methodValue
    }

    override fun visit(expression: Unary, context: InterpretationContext): Value {
        val v = eval(expression = expression.expression, context = context)
        return when (expression.operator) {
            UnaryOperator.NEG -> {
                v as Value.IntValue
                Value.IntValue(value = -v.value)
            }
            UnaryOperator.NOT -> {
                v as Value.BoolValue
                Value.BoolValue(value = !v.value)
            }
        }
    }

    override fun visit(expression: Panic, context: InterpretationContext): Nothing =
        throw PanicException(
            reason = (eval(
                expression = expression.expression,
                context = context
            ) as Value.StringValue).value
        )

    override fun visit(expression: FunctionApplication, context: InterpretationContext): Value {
        val (args, body, ctx) = eval(
            expression = expression.functionExpression,
            context = context
        ) as Value.FunctionValue
        val argValues = expression.arguments.map { eval(expression = it, context = context) }
        val bodyContext = ctx.copy(localValues = ctx.localValues.plus(pairs = args.zip(argValues)))
        return eval(expression = body, context = bodyContext)
    }

    override fun visit(expression: Binary, context: InterpretationContext): Value {
        val v1 = eval(expression = expression.e1, context = context)
        val v2 = eval(expression = expression.e2, context = context)
        return when (expression.operator) {
            MUL -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.IntValue(value = v1.value * v2.value)
            }
            DIV -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                if (v2.value == 0L) {
                    throw PanicException(reason = "Division by zero!")
                }
                Value.IntValue(value = v1.value / v2.value)
            }
            MOD -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                if (v2.value == 0L) {
                    throw PanicException(reason = "Mod by zero!")
                }
                Value.IntValue(value = v1.value % v2.value)
            }
            PLUS -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.IntValue(value = v1.value + v2.value)
            }
            MINUS -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.IntValue(value = v1.value - v2.value)
            }
            LT -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.BoolValue(value = v1.value < v2.value)
            }
            LE -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.BoolValue(value = v1.value <= v2.value)
            }
            GT -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.BoolValue(value = v1.value > v2.value)
            }
            GE -> {
                v1 as Value.IntValue
                v2 as Value.IntValue
                Value.BoolValue(value = v1.value >= v2.value)
            }
            EQ -> {
                if (v1 is Value.FunctionValue || v2 is Value.FunctionValue) {
                    throw PanicException(reason = "Cannot compare functions!")
                }
                Value.BoolValue(value = v1 == v2)
            }
            NE -> {
                if (v1 is Value.FunctionValue || v2 is Value.FunctionValue) {
                    throw PanicException(reason = "Cannot compare functions!")
                }
                Value.BoolValue(value = v1 != v2)
            }
            AND -> {
                v1 as Value.BoolValue
                v2 as Value.BoolValue
                Value.BoolValue(value = v1.value && v2.value)
            }
            OR -> {
                v1 as Value.BoolValue
                v2 as Value.BoolValue
                Value.BoolValue(value = v1.value || v2.value)
            }
        }
    }

    override fun visit(expression: IfElse, context: InterpretationContext): Value = eval(
        expression = if ((eval(
                expression = expression.boolExpression,
                context = context
            ) as Value.BoolValue).value
        ) expression.e1 else expression.e2,
        context = context
    )

    override fun visit(expression: Match, context: InterpretationContext): Value {
        val matchedValue = eval(expression = expression.matchedExpression, context = context) as Value.VariantValue
        val matchedPattern = expression.matchingList.find { it.tag == matchedValue.tag } ?: blameTypeChecker()
        val ctx = matchedPattern.dataVariable?.let { variable ->
            context.copy(localValues = context.localValues.plus(pair = variable to matchedValue.data))
        } ?: context
        return eval(expression = matchedPattern.expression, context = ctx)
    }

    override fun visit(expression: Lambda, context: InterpretationContext): Value.FunctionValue = Value.FunctionValue(
        arguments = expression.parameters.map { it.first },
        body = expression.body,
        context = context
    )

    override fun visit(expression: Val, context: InterpretationContext): Value {
        val assignedValue = eval(expression = expression.assignedExpression, context = context)
        val nextExpression = expression.nextExpression
        val ctx = when (val p = expression.pattern) {
            is Pattern.TuplePattern -> {
                val tupleValues = (assignedValue as Value.TupleValue).tupleContent
                val additionalMappings = p.destructedNames.zip(tupleValues).mapNotNull { (name, value) ->
                    name?.let { it to value }
                }
                context.copy(localValues = context.localValues.plus(pairs = additionalMappings))
            }
            is Pattern.ObjectPattern -> {
                val objectValueMappings = (assignedValue as Value.ObjectValue).objectContent
                val additionalMappings = p.destructedNames.map { (original, renamed) ->
                    val v = objectValueMappings[original] ?: blameTypeChecker()
                    (renamed ?: original) to v
                }
                context.copy(localValues = context.localValues.plus(pairs = additionalMappings))
            }
            is Pattern.VariablePattern -> context.copy(
                localValues = context.localValues.plus(pair = p.name to assignedValue)
            )
            is Pattern.WildCardPattern -> context
        }
        if (nextExpression == null) {
            return Value.UnitValue
        }
        return eval(expression = nextExpression, context = ctx)
    }
}

File: typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts
import { IdentifierType } from '../ast/common/types';
import { SamlangExpression, EXPRESSION_VARIABLE } from '../ast/lang/samlang-expressions';
import { InterpretationContext, EMPTY } from './interpretation-context';
import PanicException from './panic-exception';
import { Value, ObjectValue, FunctionValue, VariantValue, TupleValue } from './value';

export default class ExpressionInterpreter {
  private printedCollector = '';

  private blameTypeChecker = (message = ''): never => {
    throw Error(message);
  };

  readonly eval = (
    expression: SamlangExpression,
    context: InterpretationContext = EMPTY
  ): Value => {
    switch (expression.__type__) {
      case 'LiteralExpression': {
        switch (expression.literal.type) {
          case 'IntLiteral':
            return BigInt(expression.literal.value);
          case 'StringLiteral':
          case 'BoolLiteral':
            return expression.literal.value;
        }
      }
      // eslint-disable-next-line no-fallthrough
      case 'ThisExpression':
        return context.localValues.this ?? this.blameTypeChecker('Missing `this`');
      case 'VariableExpression':
        return (
          context.localValues[expression.name] ??
          this.blameTypeChecker(`Missing variable ${expression.name}`)
        );
      case 'ClassMemberExpression':
        return (
          context.classes[expression.className]?.functions?.[expression.memberName] ||
          this.blameTypeChecker()
        );
      case 'TupleConstructorExpression':
        return {
          type: 'tuple',
          tupleContent: expression.expressions.map((e) => this.eval(e, context)),
        };
      case 'ObjectConstructorExpression': {
        const objectContent = new Map<string, Value>();
        expression.fieldDeclarations.forEach((declaration) => {
          if (declaration.expression) {
            objectContent.set(declaration.name, this.eval(declaration.expression, context));
          } else {
            const { range, type, name } = declaration;
            objectContent.set(
              declaration.name,
              this.eval(EXPRESSION_VARIABLE({ range, type, name }), context)
            );
          }
        });
        return { type: 'object', objectContent };
      }
      case 'VariantConstructorExpression':
        return { type: 'variant', tag: expression.tag, data: this.eval(expression.data, context) };
      case 'FieldAccessExpression': {
        const thisValue = this.eval(expression.expression, context) as ObjectValue;
        return thisValue.objectContent?.get(expression.fieldName) ?? this.blameTypeChecker();
      }
      case 'MethodAccessExpression': {
        const identifier = (expression.expression.type as IdentifierType).identifier;
        const thisValue = this.eval(expression.expression, context);
        const methodValue =
          context.classes[identifier]?.methods?.[expression.methodName] ?? this.blameTypeChecker();
        methodValue.context = {
          classes: context.classes,
          localValues: { ...context.localValues, this: thisValue },
        };
        return methodValue;
      }
      case 'UnaryExpression': {
        const v = this.eval(expression.expression);
        switch (expression.operator) {
          case '-':
            return -v;
          case '!':
            return !v;
        }
      }
      // eslint-disable-next-line no-fallthrough
      case 'PanicExpression':
        throw new PanicException(this.eval(expression.expression, context) as string);
      case 'BuiltInFunctionCallExpression': {
        const argumentValue = this.eval(expression.argumentExpression, context);
        switch (expression.functionName) {
          case 'stringToInt': {
            const value = argumentValue as string;
            const parsedValue = parseInt(value, 10);
            if (!Number.isNaN(parsedValue)) {
              return BigInt(parsedValue);
            }
            throw new PanicException(`Cannot convert \`${value}\` to int.`);
          }
          case 'intToString':
            return (argumentValue as bigint).toString();
          case 'println':
            this.printedCollector.concat(`${argumentValue as string}\n`);
            return { type: 'unit' };
        }
      }
      // eslint-disable-next-line no-fallthrough
      case 'FunctionCallExpression': {
        const functionVal = this.eval(expression.functionExpression) as FunctionValue;
        const args = functionVal.arguments;
        const body = functionVal.body;
        const ctx = functionVal.context;
        const argValues = expression.functionArguments.map((arg) => this.eval(arg, context));
        const bodyLocalValues = { ...ctx.localValues };
        args.forEach((arg, i) => {
          bodyLocalValues[arg] = argValues[i];
        });
        const bodyContext = { classes: ctx.classes, localValues: { ...bodyLocalValues } };
        return this.eval(body, bodyContext);
      }
      case 'BinaryExpression': {
        switch (expression.operator.symbol) {
          case '*': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return BigInt(v1 * v2);
          }
          case '/': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            if (v2 === BigInt(0)) {
              throw new PanicException('Division by zero!');
            }
            return BigInt(v1 / v2);
          }
          case '%': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            if (v2 === BigInt(0)) {
              throw new PanicException('Mod by zero!');
            }
            return BigInt(v1 % v2);
          }
          case '+': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return BigInt(v1 + v2);
          }
          case '-': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return BigInt(v1 - v2);
          }
          case '<': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return v1 < v2;
          }
          case '<=': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return v1 <= v2;
          }
          case '>': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return v1 > v2;
          }
          case '>=': {
            const v1 = this.eval(expression.e1) as bigint;
            const v2 = this.eval(expression.e2) as bigint;
            return v1 >= v2;
          }
          case '==': {
            const v1 = this.eval(expression.e1);
            const v2 = this.eval(expression.e2);
            if (
              (v1 as FunctionValue).type === 'functionValue' ||
              (v2 as FunctionValue).type === 'functionValue'
            ) {
              throw new PanicException('Cannot compare functions!');
            }
            return v1 === v2;
          }
          case '!=': {
            const v1 = this.eval(expression.e1);
            const v2 = this.eval(expression.e2);
            if (
              (v1 as FunctionValue).type === 'functionValue' ||
              (v2 as FunctionValue).type === 'functionValue'
            ) {
              throw new PanicException('Cannot compare functions!');
            }
            return v1 !== v2;
          }
          case '&&': {
            const v1 = this.eval(expression.e1) as boolean;
            return v1 && this.eval(expression.e2, context);
          }
          case '||': {
            const v1 = this.eval(expression.e1) as boolean;
            return v1 || this.eval(expression.e2, context);
          }
          case '::': {
            const v1 = this.eval(expression.e1) as string;
            const v2 = this.eval(expression.e2) as string;
            return v1 + v2;
          }
        }
      }
      // eslint-disable-next-line no-fallthrough
      case 'IfElseExpression': {
        return this.eval(
          (this.eval(expression.boolExpression) as boolean) ? expression.e1 : expression.e2,
          context
        );
      }
      case 'MatchExpression': {
        const matchedValue = this.eval(expression.matchedExpression, context) as VariantValue;
        const matchedPattern =
          expression.matchingList.find((el) => el.tag === matchedValue.tag) ??
          this.blameTypeChecker();
        let ctx = context;
        if (matchedPattern.dataVariable) {
          ctx = {
            classes: ctx.classes,
            localValues: { ...ctx.localValues, [matchedPattern.dataVariable]: matchedValue.data },
          };
        }
        return this.eval(matchedPattern.expression, ctx);
      }
      case 'LambdaExpression':
        return {
          type: 'functionValue',
          arguments: expression.parameters.map((param) => param[0]),
          body: expression.body,
          context,
        };
      case 'StatementBlockExpression': {
        const { block } = expression;
        const contextForStatementBlock = { ...context, localValues: { ...context.localValues } };
        block.statements.forEach((statement) => {
          const assignedValue = this.eval(statement.assignedExpression, contextForStatementBlock);
          const p = statement.pattern;
          switch (p.type) {
            case 'TuplePattern': {
              const { tupleContent } = assignedValue as TupleValue;
              p.destructedNames.forEach((nameWithRange, i) => {
                if (nameWithRange[0] !== null) {
                  contextForStatementBlock.localValues[nameWithRange[0]] = tupleContent[i];
                }
              });
              break;
            }
            case 'ObjectPattern': {
              const { objectContent } = assignedValue as ObjectValue;
              p.destructedNames.forEach(({ fieldName, alias }) => {
                const v = objectContent.get(fieldName) ?? this.blameTypeChecker();
                contextForStatementBlock.localValues[alias ?? fieldName] = v;
              });
              break;
            }
            case 'VariablePattern':
              contextForStatementBlock.localValues[p.name] = assignedValue;
              break;
            case 'WildCardPattern':
              break;
          }
        });
        const finalExpression = block.expression;
        return finalExpression === undefined
          ? { type: 'unit' }
          : this.eval(finalExpression, contextForStatementBlock);
      }
    }
  };
}
